### PR TITLE
Fix exception in background metrics collection

### PR DESCRIPTION
**Pull request by erikjohnston**
We attempted to iterate through a list on a separate thread without
doing the necessary copying.

### CODE REVIEW

**Comment by erikjohnston** [comment]
(We probably can clean some of this exception handling up a bit more so that we don't log stack traces simply because we failed to connect to the remote server, but I'm punting that for now)

File: synapse/metrics/background_process_metrics.py [hunk 1]
```diff
@@ -101,9 +101,13 @@ def collect(self):
...
         )
 
-        # We copy the dict so that it doesn't change from underneath us
+        # We copy the dict so that it doesn't change from underneath us.
+        # We also copy the process lists as that can also change
         with _bg_metrics_lock:
-            _background_processes_copy = dict(_background_processes)
+            _background_processes_copy = {
+                k: list(v)
+                for k, v in six.iteritems(_background_processes)
```
Comment (richvdh) [inline synapse/metrics/background_process_metrics.py]: what is to say that this will be thread-safe when the loop below was not?

**Comment by erikjohnston** [comment]
```
2018-10-03 11:43:23,185 - twisted - 243 - ERROR -  - Traceback (most recent call last):
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -   File "/usr/lib/python2.7/SocketServer.py", line 596, in process_request_thread
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -     self.finish_request(request, client_address)
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -   File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -     self.RequestHandlerClass(request, client_address, self)
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -   File "/usr/lib/python2.7/SocketServer.py", line 652, in __init__
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -     self.handle()
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -   File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -     self.handle_one_request()
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -   File "/usr/lib/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
2018-10-03 11:43:23,186 - twisted - 243 - ERROR -  -     method()
2018-10-03 11:43:23,187 - twisted - 243 - ERROR -  -   File "/home/matrix/.synapse/local/lib/python2.7/site-packages/prometheus_client/exposition.py", line 95, in do_GET
2018-10-03 11:43:23,187 - twisted - 243 - ERROR -  -     output = generate_latest(registry)
2018-10-03 11:43:23,187 - twisted - 243 - ERROR -  -   File "/home/matrix/.synapse/local/lib/python2.7/site-packages/prometheus_client/exposition.py", line 69, in generate_latest
2018-10-03 11:43:23,187 - twisted - 243 - ERROR -  -     for metric in registry.collect():
2018-10-03 11:43:23,187 - twisted - 243 - ERROR -  -   File "synapse/metrics/__init__.py", line 46, in collect
2018-10-03 11:43:23,187 - twisted - 243 - ERROR -  -   File "/home/matrix/.synapse/local/lib/python2.7/site-packages/prometheus_client/core.py", line 102, in collect
2018-10-03 11:43:23,188 - twisted - 243 - ERROR -  -     for metric in collector.collect():
2018-10-03 11:43:23,188 - twisted - 243 - ERROR -  -   File "synapse/metrics/background_process_metrics.py", line 112, in collect
2018-10-03 11:43:23,188 - twisted - 243 - ERROR -  - RuntimeError: Set changed size during iteration
```

File: synapse/metrics/background_process_metrics.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/metrics/background_process_metrics.py]: We're copying the dict and lists from inside a lock

Comment (richvdh) [reply synapse/metrics/background_process_metrics.py]: ffs. I should probably learn to read.

Comment (erikjohnston) [reply synapse/metrics/background_process_metrics.py]: Glad I'm not the only one :grinning:

**Pull request merged**